### ISSUE TITLE
Metadata for discovered models?

### ISSUE THREAD

**Issue by rjmholt** [parsonsmatt/persistent-discover#12]
Currently, from what I can tell, the persistent-discover function (`discoverEntites` is maybe the particular Haskell implementation) is one-way in the sense that information is lost about the file metadata for those models. So any analysis or transformation of the `EntityDef`s from those files can't be related back. This data would be really useful to have for tooling around linting and generation (I'm currently trying to write a generator for history tables).

So basically, I'd be be happy with a filepath on each entitydef, although I think ideally it'd be something like:
- An extent: The filepath, lines and columns
- When the entity was parsed
- Maybe extra metadata that would be helpful for tooling persistent that I'm not thinking about or don't know enough about persistent to see the need for?

**Comment by parsonsmatt** [comment]
Hmm - that information isn't really lost, it's just never available in the first place. There are two things that happen here:

1. The `pgmF` formatter makes a big `import` list of every module in the subdirectory
2. There's a call to `discoverEntities` that calls `reifyInstances` to get the instances of `PersistEntity` into scope.

The `EntityDef` contains enough information to make a `PersistEntity` instance for it, so you could in theory have a function `mkHistoryTable :: EntityDef -> Q [Dec]` that converts the given definition into the splices rquired for a `HistoryTable` on that definition.

You can use `Typeable` to get a `TypeRep`, which has module definition information. I think you can [`discoverInstances`](https://hackage.haskell.org/package/discover-instances-0.1.0.0/docs/DiscoverInstances.html) that by making a class:

```haskell
class (Typeable a, PersistEntity a) => TypeableEntity a
instance (Typeable a, PersistEntity a) => TypeableEntity a

typeableEntities :: [SomeDict TypeableEntity]
typeableEntities = $$(discoverInstances)
```

But I think I'd need to know more about what you want to do with the information you're requesting. Some of it isn't generally available to any part of the `persistent` process (ie true line or column numbers - you know it for a file, but not for a quasiquoter)

**Comment by rjmholt** [comment]
Ah ok, it sounds like what I'm asking for isn't available at this layer — will close
